File: javascript/leetcode/balanced-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {
    function checkBalance(root) {
        if (root === null) {
            return 0;
        }
        var leftBalance = checkBalance(root.left);
        var rightBalance = checkBalance(root.right);
        if (leftBalance === -1 || rightBalance === -1 || Math.abs(leftBalance - rightBalance) > 1) {
            return -1;
        }
        return 1 + Math.max(leftBalance, rightBalance);
    }
    return checkBalance(root) !== -1;
};

File: java/leetcode/balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        return checkBalance(root) != -1;
    }
    
    private int checkBalance(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int leftBalance = checkBalance(root.left);
        int rightBalance = checkBalance(root.right);
        if (leftBalance == -1 || rightBalance == -1 || Math.abs(leftBalance - rightBalance) > 1) {
            return -1;
        }
        return 1 + Math.max(leftBalance, rightBalance);
    }
}
